node/store.go

```diff
@@ -318,6 +398,147 @@ func (s *Store) StoreBatch(ctx context.Context, header *core.BatchHeader, blobs
+
+	// Setting the expiration time for the batch.
+	curr := time.Now().Unix()
+	timeToExpire := (s.blockStaleMeasure + s.storeDurationBlocks) * 12 // 12s per block
+	// Why this expiration time is safe?
+	//
+	// The batch must be confirmed before referenceBlockNumber+blockStaleMeasure, otherwise
+	// it's stale and won't be accepted onchain. This means the blob's lifecycle will end
+	// before referenceBlockNumber+blockStaleMeasure+storeDurationBlocks.
+	// Since time@referenceBlockNumber < time.Now() (we always use a reference block that's
+	// already onchain), we have
+	// time@(referenceBlockNumber+blockStaleMeasure+storeDurationBlocks)
+	// = time@referenceBlockNumber + 12*(blockStaleMeasure+storeDurationBlocks)
+	// < time.Now() + 12*(blockStaleMeasure+storeDurationBlocks).
+	//
+	// Note if a batch is unconfirmed, it could be removed even earlier; here we treat its
+	// lifecycle the same as confirmed batches for simplicity.
+	expirationTime := curr + int64(timeToExpire)
+	expirationKey := EncodeBlobExpirationKey(expirationTime)
+	// expirationValue is a list of blob header hashes that are expired.
+	expirationValue := make([]byte, 0)
+	var err error
+	// If there is already an expiration key in the store, we need to get the value and append to it.
+	if s.HasKey(ctx, expirationKey) {
+		expirationValue, err = s.db.Get(expirationKey)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get the expiration value: %w", err)
+		}
+	}
+
+	// Generate key/value pairs for all blob headers and blob chunks .
+	size := int64(0)
+	var serializationDuration, encodingDuration time.Duration
+	for idx, blob := range blobs {
+		rawBlob := blobsProto[idx]
+		if len(rawBlob.GetBundles()) != len(blob.Bundles) {
+			return nil, fmt.Errorf("internal error: the number of bundles in parsed blob (%d) must be the same as in raw blob (%d)", len(rawBlob.GetBundles()), len(blob.Bundles))
+		}
+
+		// blob header
+		blobHeaderHash, err := blob.BlobHeader.GetBlobHeaderHash()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get blob header hash: %w", err)
+		}
+		blobHeaderKey := EncodeBlobHeaderKeyByHash(blobHeaderHash)
+
+		if s.HasKey(ctx, blobHeaderKey) {
+			s.logger.Warn("Blob already exists", "blobHeaderHash", hexutil.Encode(blobHeaderHash[:]))
+			continue
+		}
+
+		blobHeaderBytes, err := proto.Marshal(blobsProto[idx].GetHeader())
+		if err != nil {
+			return nil, fmt.Errorf("failed to serialize the blob header proto: %w", err)
+		}
+		keys = append(keys, blobHeaderKey)
+		values = append(values, blobHeaderBytes)
+		expirationValue = append(expirationValue, blobHeaderHash[:]...)
+
+		// Get raw chunks
+		start := time.Now()
+		format := GetBundleEncodingFormat(rawBlob)
+		rawBundles := make(map[core.QuorumID][]byte)
+		rawChunks := make(map[core.QuorumID][][]byte)
+		for i, bundle := range rawBlob.GetBundles() {
+			quorumID := uint8(rawBlob.GetHeader().GetQuorumHeaders()[i].GetQuorumId())
+			if format == core.GnarkBundleEncodingFormat {
```

Comment:
It seems unideal that we are processing the chunk format both here at when we process the raw blobs in `GetBlobMessages` of `utils.go`. I'd prefer it if `GetBlobMessages` or some other utility function did all of this processing and then returned all formats of the chunks needed by validation and storage, so that the storage functions can be format agnostic.

Reply:
Agree it's a bit agree. I was mostly leaving the implementation same as `StoreBatch` to minimize risk
On the bright side, this branching based on encoding format will also go away soon when all blobs are encoded with gnark